### DISCUSSION TITLE
Laravel Spark

### DISCUSSION THREAD

**Discussion by masterix21** [Laravel-Lang/common#38] [Q&A]
Hi team, what do you think about Laravel Spark? Should we add it? @andrey-helldar @caouecs 

This is the `en.json` file:
```json
{
    ":days day trial" : ":days day trial",
    "Add VAT Number" : "Add VAT Number",
    "Address Line 2" : "Address Line 2",
    "Address" : "Address",
    "Afghanistan" : "Afghanistan",
    "Albania" : "Albania",
    "Algeria" : "Algeria",
    "American Samoa" : "American Samoa",
    "An unexpected error occurred and we have notified our support team. Please try again later." : "An unexpected error occurred and we have notified our support team. Please try again later.",
    "Andorra" : "Andorra",
    "Angola" : "Angola",
    "Anguilla" : "Anguilla",
    "Antarctica" : "Antarctica",
    "Antigua and Barbuda" : "Antigua and Barbuda",
    "Argentina" : "Argentina",
    "Armenia" : "Armenia",
    "Aruba" : "Aruba",
    "Australia" : "Australia",
    "Austria" : "Austria",
    "Azerbaijan" : "Azerbaijan",
    "Bahamas" : "Bahamas",
    "Bahrain" : "Bahrain",
    "Bangladesh" : "Bangladesh",
    "Barbados" : "Barbados",
    "Belarus" : "Belarus",
    "Belgium" : "Belgium",
    "Belize" : "Belize",
    "Benin" : "Benin",
    "Bermuda" : "Bermuda",
    "Bhutan" : "Bhutan",
    "Billing Information" : "Billing Information",
    "Billing Management" : "Billing Management",
    "Bolivia, Plurinational State of" : "Bolivia, Plurinational State of",
    "Bosnia and Herzegovina" : "Bosnia and Herzegovina",
    "Botswana" : "Botswana",
    "Bouvet Island" : "Bouvet Island",
    "Brazil" : "Brazil",
    "British Indian Ocean Territory" : "British Indian Ocean Territory",
    "Brunei Darussalam" : "Brunei Darussalam",
    "Bulgaria" : "Bulgaria",
    "Burkina Faso" : "Burkina Faso",
    "Burundi" : "Burundi",
    "Cambodia" : "Cambodia",
    "Cameroon" : "Cameroon",
    "Canada" : "Canada",
    "Cancel Subscription" : "Cancel Subscription",
    "Cape Verde" : "Cape Verde",
    "Card" : "Card",
    "Cayman Islands" : "Cayman Islands",
    "Central African Republic" : "Central African Republic",
    "Chad" : "Chad",
    "Change Subscription Plan" : "Change Subscription Plan",
    "Chile" : "Chile",
    "China" : "China",
    "Christmas Island" : "Christmas Island",
    "City" : "City",
    "Cocos (Keeling) Islands" : "Cocos (Keeling) Islands",
    "Colombia" : "Colombia",
    "Comoros" : "Comoros",
    "Confirm Payment" : "Confirm Payment",
    "Confirm your :amount payment" : "Confirm your :amount payment",
    "Congo" : "Congo",
    "Congo, the Democratic Republic of the" : "Congo, the Democratic Republic of the",
    "Cook Islands" : "Cook Islands",
    "Costa Rica" : "Costa Rica",
    "Country" : "Country",
    "Coupon" : "Coupon",
    "Croatia" : "Croatia",
    "Cuba" : "Cuba",
    "Current Subscription Plan" : "Current Subscription Plan",
    "Currently Subscribed" : "Currently Subscribed",
    "Cyprus" : "Cyprus",
    "Czech Republic" : "Czech Republic",
    "Côte d'Ivoire" : "Côte d'Ivoire",
    "Denmark" : "Denmark",
    "Djibouti" : "Djibouti",
    "Dominica" : "Dominica",
    "Dominican Republic" : "Dominican Republic",
    "Download Receipt" : "Download Receipt",
    "ex VAT" : "ex VAT",
    "Ecuador" : "Ecuador",
    "Egypt" : "Egypt",
    "El Salvador" : "El Salvador",
    "Equatorial Guinea" : "Equatorial Guinea",
    "Eritrea" : "Eritrea",
    "Estonia" : "Estonia",
    "Ethiopia" : "Ethiopia",
    "Email Addresses" : "Email Addresses",
    "Extra Billing Information" : "Extra Billing Information",
    "Extra confirmation is needed to process your payment. Please continue to the payment page by clicking on the button below." : "Extra confirmation is needed to process your payment. Please continue to the payment page by clicking on the button below.",
    "Falkland Islands (Malvinas)" : "Falkland Islands (Malvinas)",
    "Faroe Islands" : "Faroe Islands",
    "Fiji" : "Fiji",
    "Finland" : "Finland",
    "France" : "France",
    "French Guiana" : "French Guiana",
    "French Polynesia" : "French Polynesia",
    "French Southern Territories" : "French Southern Territories",
    "Gabon" : "Gabon",
    "Gambia" : "Gambia",
    "Georgia" : "Georgia",
    "Germany" : "Germany",
    "Ghana" : "Ghana",
    "Gibraltar" : "Gibraltar",
    "Greece" : "Greece",
    "Greenland" : "Greenland",
    "Grenada" : "Grenada",
    "Guadeloupe" : "Guadeloupe",
    "Guam" : "Guam",
    "Guatemala" : "Guatemala",
    "Guernsey" : "Guernsey",
    "Guinea" : "Guinea",
    "Guinea-Bissau" : "Guinea-Bissau",
    "Guyana" : "Guyana",
    "Haiti" : "Haiti",
    "Have a coupon code?" : "Have a coupon code?",
    "Having second thoughts about cancelling your subscription? You can instantly reactive your subscription at any time until the end of your current billing cycle. After your current billing cycle ends, you may choose an entirely new subscription plan." : "Having second thoughts about cancelling your subscription? You can instantly reactive your subscription at any time until the end of your current billing cycle. After your current billing cycle ends, you may choose an entirely new subscription plan.",
    "Heard Island and McDonald Islands" : "Heard Island and McDonald Islands",
    "Holy See (Vatican City State)" : "Holy See (Vatican City State)",
    "Honduras" : "Honduras",
    "Hong Kong" : "Hong Kong",
    "Hungary" : "Hungary",
    "Iceland" : "Iceland",
    "If you need to add specific contact or tax information to your receipts, like your full business name, VAT identification number, or address of record, you may add it here." : "If you need to add specific contact or tax information to your receipts, like your full business name, VAT identification number, or address of record, you may add it here.",
    "India" : "India",
    "Indonesia" : "Indonesia",
    "Iran, Islamic Republic of" : "Iran, Islamic Republic of",
    "Iraq" : "Iraq",
    "Ireland" : "Ireland",
    "Isle of Man" : "Isle of Man",
    "Israel" : "Israel",
    "It looks like you do not have an active subscription. You may choose one of the subscription plans below to get started. Subscription plans may be changed or cancelled at your convenience." : "It looks like you do not have an active subscription. You may choose one of the subscription plans below to get started. Subscription plans may be changed or cancelled at your convenience.",
    "Italy" : "Italy",
    "Jamaica" : "Jamaica",
    "Japan" : "Japan",
    "Jersey" : "Jersey",
    "Jordan" : "Jordan",
    "Kazakhstan" : "Kazakhstan",
    "Kenya" : "Kenya",
    "Kiribati" : "Kiribati",
    "Korea, Democratic People's Republic of" : "Korea, Democratic People's Republic of",
    "Korea, Republic of" : "Korea, Republic of",
    "Kuwait" : "Kuwait",
    "Kyrgyzstan" : "Kyrgyzstan",
    "Lao People's Democratic Republic" : "Lao People's Democratic Republic",
    "Latvia" : "Latvia",
    "Lebanon" : "Lebanon",
    "Lesotho" : "Lesotho",
    "Liberia" : "Liberia",
    "Libyan Arab Jamahiriya" : "Libyan Arab Jamahiriya",
    "Liechtenstein" : "Liechtenstein",
    "Lithuania" : "Lithuania",
    "Luxembourg" : "Luxembourg",
    "Macao" : "Macao",
    "Macedonia, the former Yugoslav Republic of" : "Macedonia, the former Yugoslav Republic of",
    "Madagascar" : "Madagascar",
    "Malawi" : "Malawi",
    "Malaysia" : "Malaysia",
    "Maldives" : "Maldives",
    "Mali" : "Mali",
    "Malta" : "Malta",
    "Managing billing for :billableName" : "Managing billing for :billableName",
    "Marshall Islands" : "Marshall Islands",
    "Martinique" : "Martinique",
    "Mauritania" : "Mauritania",
    "Mauritius" : "Mauritius",
    "Mayotte" : "Mayotte",
    "Mexico" : "Mexico",
    "Micronesia, Federated States of" : "Micronesia, Federated States of",
    "Moldova, Republic of" : "Moldova, Republic of",
    "Monaco" : "Monaco",
    "Mongolia" : "Mongolia",
    "Montenegro" : "Montenegro",
    "Monthly" : "Monthly",
    "monthly" : "monthly",
    "Montserrat" : "Montserrat",
    "Morocco" : "Morocco",
    "Mozambique" : "Mozambique",
    "Myanmar" : "Myanmar",
    "Namibia" : "Namibia",
    "Nauru" : "Nauru",
    "Nepal" : "Nepal",
    "Netherlands Antilles" : "Netherlands Antilles",
    "Netherlands" : "Netherlands",
    "Nevermind, I'll keep my old plan" : "Nevermind, I'll keep my old plan",
    "New Caledonia" : "New Caledonia",
    "New Zealand" : "New Zealand",
    "Nicaragua" : "Nicaragua",
    "Niger" : "Niger",
    "Nigeria" : "Nigeria",
    "Niue" : "Niue",
    "Norfolk Island" : "Norfolk Island",
    "Northern Mariana Islands" : "Northern Mariana Islands",
    "Norway" : "Norway",
    "Oman" : "Oman",
    "Our billing management portal allows you to conveniently manage your subscription plan, payment method, and download your recent invoices." : "Our billing management portal allows you to conveniently manage your subscription plan, payment method, and download your recent invoices.",
    "Pakistan" : "Pakistan",
    "Palau" : "Palau",
    "Palestinian Territory, Occupied" : "Palestinian Territory, Occupied",
    "Panama" : "Panama",
    "Papua New Guinea" : "Papua New Guinea",
    "Paraguay" : "Paraguay",
    "Payment Information" : "Payment Information",
    "Peru" : "Peru",
    "Philippines" : "Philippines",
    "Pitcairn" : "Pitcairn",
    "Poland" : "Poland",
    "Portugal" : "Portugal",
    "Puerto Rico" : "Puerto Rico",
    "Please provide a maximum of three receipt emails addresses.": "Please provide a maximum of three receipt emails addresses.",
    "Qatar" : "Qatar",
    "Receipts" : "Receipts",
    "Receipt Email Addresses" : "Receipt Email Addresses",
    "Resume Subscription" : "Resume Subscription",
    "Return to :appName" : "Return to :appName",
    "Romania" : "Romania",
    "Russian Federation" : "Russian Federation",
    "Rwanda" : "Rwanda",
    "Réunion" : "Réunion",
    "Saint Barthélemy" : "Saint Barthélemy",
    "Saint Helena" : "Saint Helena",
    "Saint Kitts and Nevis" : "Saint Kitts and Nevis",
    "Saint Lucia" : "Saint Lucia",
    "Saint Martin (French part)" : "Saint Martin (French part)",
    "Saint Pierre and Miquelon" : "Saint Pierre and Miquelon",
    "Saint Vincent and the Grenadines" : "Saint Vincent and the Grenadines",
    "Samoa" : "Samoa",
    "San Marino" : "San Marino",
    "Sao Tome and Principe" : "Sao Tome and Principe",
    "Saudi Arabia" : "Saudi Arabia",
    "Save": "Save",
    "Select a different plan" : "Select a different plan",
    "Select" : "Select",
    "Senegal" : "Senegal",
    "Serbia" : "Serbia",
    "Seychelles" : "Seychelles",
    "Sierra Leone" : "Sierra Leone",
    "Signed in as" : "Signed in as",
    "Singapore" : "Singapore",
    "Slovakia" : "Slovakia",
    "Slovenia" : "Slovenia",
    "Solomon Islands" : "Solomon Islands",
    "Somalia" : "Somalia",
    "South Africa" : "South Africa",
    "South Georgia and the South Sandwich Islands" : "South Georgia and the South Sandwich Islands",
    "Spain" : "Spain",
    "Sri Lanka" : "Sri Lanka",
    "State / County" : "State / County",
    "Subscribe" : "Subscribe",
    "Subscription Information" : "Subscription Information",
    "Sudan" : "Sudan",
    "Suriname" : "Suriname",
    "Svalbard and Jan Mayen" : "Svalbard and Jan Mayen",
    "Swaziland" : "Swaziland",
    "Sweden" : "Sweden",
    "Switzerland" : "Switzerland",
    "Syrian Arab Republic" : "Syrian Arab Republic",
    "Taiwan, Province of China" : "Taiwan, Province of China",
    "Tajikistan" : "Tajikistan",
    "Tanzania, United Republic of" : "Tanzania, United Republic of",
    "Thailand" : "Thailand",
    "Thanks for your continued support. We've attached a copy of your invoice for your records. Please let us know if you have any questions or concerns." : "Thanks for your continued support. We've attached a copy of your invoice for your records. Please let us know if you have any questions or concerns.",
    "Thanks," : "Thanks,",
    "The provided VAT number is invalid." : "The provided VAT number is invalid.",
    "The provided coupon code is invalid." : "The provided coupon code is invalid.",
    "The selected country is invalid." : "The selected country is invalid.",
    "The selected plan is invalid." : "The selected plan is invalid.",
    "Terms of Service" : "Terms of Service",
    "This account does not have an active subscription." : "This account does not have an active subscription.",
    "The receipt emails must be valid email addresses.": "The receipt emails must be valid email addresses.",
    "This subscription has expired and cannot be resumed. Please create a new subscription." : "This subscription has expired and cannot be resumed. Please create a new subscription.",
    "Timor-Leste" : "Timor-Leste",
    "Togo" : "Togo",
    "Tokelau" : "Tokelau",
    "Tonga" : "Tonga",
    "Total:" : "Total:",
    "Trinidad and Tobago" : "Trinidad and Tobago",
    "Tunisia" : "Tunisia",
    "Turkey" : "Turkey",
    "Turkmenistan" : "Turkmenistan",
    "Turks and Caicos Islands" : "Turks and Caicos Islands",
    "Tuvalu" : "Tuvalu",
    "Uganda" : "Uganda",
    "Ukraine" : "Ukraine",
    "United Arab Emirates" : "United Arab Emirates",
    "United Kingdom" : "United Kingdom",
    "United States Minor Outlying Islands" : "United States Minor Outlying Islands",
    "United States" : "United States",
    "Update Payment Information" : "Update Payment Information",
    "Update" : "Update",
    "Uruguay" : "Uruguay",
    "Uzbekistan" : "Uzbekistan",
    "VAT Number" : "VAT Number",
    "Vanuatu" : "Vanuatu",
    "Venezuela, Bolivarian Republic of" : "Venezuela, Bolivarian Republic of",
    "Viet Nam" : "Viet Nam",
    "Virgin Islands, British" : "Virgin Islands, British",
    "Virgin Islands, U.S." : "Virgin Islands, U.S.",
    "Wallis and Futuna" : "Wallis and Futuna",
    "We are unable to process your payment. Please contact customer support." : "We are unable to process your payment. Please contact customer support.",
    "We will send a receipt download link to the email addresses that you specify below. You may separate multiple email addresses using commas." : "We will send a receipt download link to the email addresses that you specify below. You may separate multiple email addresses using commas.",
    "Western Sahara" : "Western Sahara",
    "Whoops! Something went wrong." : "Whoops! Something went wrong.",
    "Yearly" : "Yearly",
    "Yemen" : "Yemen",
    "You are currently within your free trial period. Your trial will expire on :date." : "You are currently within your free trial period. Your trial will expire on :date.",
    "You may cancel your subscription at any time. Once your subscription has been cancelled, you will have the option to resume the subscription until the end of your current billing cycle." : "You may cancel your subscription at any time. Once your subscription has been cancelled, you will have the option to resume the subscription until the end of your current billing cycle.",
    "Your :invoiceName invoice is now available!" : "Your :invoiceName invoice is now available!",
    "Your card was declined. Please contact your card issuer for more information." : "Your card was declined. Please contact your card issuer for more information.",
    "Your current payment method is a credit card ending in :lastFour that expires on :expiration." : "Your current payment method is a credit card ending in :lastFour that expires on :expiration.",
    "Your registered VAT Number is :vatNumber." : "Your registered VAT Number is :vatNumber.",
    "Zambia" : "Zambia",
    "Zimbabwe" : "Zimbabwe",
    "Zip / Postal Code" : "Zip / Postal Code",
    "Åland Islands" : "Åland Islands"
}

```

**Comment by andrey-helldar** [comment]
Our project contains the translation of phrases used in Laravel products. So yes, I think it needs to be translated.

From the positive - the names of the countries have already been translated))

**Reply by caouecs** [reply]
If we add these tags now, we will have problem with translation from machine

**Reply by andrey-helldar** [reply]
Yes. Therefore, I suggest adding them after we have processed all the PRs with machine translation.

**Reply by caouecs** [reply]
I think in 3 days, without answer, we can push all machine translations, and if it's not good, they can do pull request to fix :)

**Comment by caouecs** [comment]
Tomorrow I will add these tags

**Reply by luisprmat** [reply]
I have started to translate the tags for Laravel Spark to spanish but before my PR i'll wait for update of  `source/en.json` to avoid conflicts unexpected.

**Comment by caouecs** [comment]
I don't know if it's interesting to have an only project with all translations from different Laravel packages.

It's just for a person who wants just the tags for Laravel Framework, he has all tags...

It's just a reflexion, we have a centralized file with all tags, and after we can cut but package

**Reply by luisprmat** [reply]
I don't seem a good idea to separate translations by package bacause there are several tags shared between packages and we should be duplicating the code.
I'd like a way to flag each translation (specially in *.json) with the packages where it's used. (Example "Cancel" is used in *laravel framework*, *laravel fortify*, *laravel jetstream*,...) and a command to install translations only of used packages (it will be amazing that this was read from _composer.json_) and also for required languages.
I know that it exists packages for install languages but i don't know if they have this feature.

**Reply by andrey-helldar** [reply]
In my opinion, splitting translation files into different packages will help track changes in them. For example, you can write a script that will use a regular expression to search for references to translation files and automatically compose a list of them.

But in this case, we will encounter the second problem - code duplication. For example, the Jetstream and Spark projects contain a list of countries, which is about 200 entries per file.

**Reply by caouecs** [reply]
In fact, we have just one json file by language, with all tags. In source directory ( or another one ) we have the list of tags by package.

And a script generates all files for packages.

**Comment by caouecs** [comment]
Just a question : Laravel Spark is free ? open source ?

**Reply by luisprmat** [reply]
It's paid

**Reply by caouecs** [reply]
So, why we translate something not free ?

**Reply by luisprmat** [reply]
I have always thought that we are translating packages that we don't even have access to, but this was already done with **Laravel Nova** (paid). it would be good as a team to set a position on this.

**Reply by caouecs** [reply]
Oups !!! for Nova, I didn't know

**Reply by luisprmat** [reply]
**Laravel Nova** and **Laravel Spark** are the only packages that have been considered for translating that are paid. The others are free.

**Reply by luisprmat** [reply]
> Oups !!! for Nova, I didn't know

![image](https://user-images.githubusercontent.com/9275870/116006209-c19c6e80-a5cf-11eb-95f3-6f35579d208e.png)

**Reply by caouecs** [reply]
I didn't follow different Laravel packages, sorry ;)

**Reply by andrey-helldar** [reply]
@caouecs, I knew that Nova was paid. I thought we were doing translation for all Laravel projects regardless of distribution type. Sorry.

**Reply by caouecs** [reply]
I'm not against translating paid package, but it's most difficult to do homogeneity between projects ( see Laravel-Lang/lang#1716 )

**Comment by caouecs** [comment]
I will try to help translators with country names ( duplication )

**Comment by masterix21** [comment]
I've found a "bug": currently, our Spark translations don't work, because the `SparkServiceProvider` loads these from the spark folder:

![image](https://user-images.githubusercontent.com/6555012/117281883-69355e80-ae64-11eb-9e89-9c6537dc1e34.png)

What can we do? @caouecs @andrey-helldar

**Reply by andrey-helldar** [reply]
Are translations not working exactly?

Now I see that Laravel Spark is given a separate path for searching for files, despite the fact that it calls the `__()` helper, which searches through all application files. Therefore, he must take values from the default paths as well.

![image](https://user-images.githubusercontent.com/10347617/117284388-6ee47180-ae6f-11eb-92e1-17f378652db7.png)

https://spark.laravel.com/docs/1.x/spark-stripe/customization.html#localization

**Reply by caouecs** [reply]
We need a json file just for Spark ?

**Reply by luisprmat** [reply]
I did not know since I do not have access to this package, but everything seems to indicate that we should have separated the translations. In this way we clean the `.json` a bit. Doesn't the same happen with Laravel Nova?

![image](https://user-images.githubusercontent.com/9275870/117310245-b3104b80-ae48-11eb-8f98-17b4ffb1ca83.png)
[https://nova.laravel.com/docs/3.0/customization/localization.html#localization](https://nova.laravel.com/docs/3.0/customization/localization.html#localization)

**Reply by andrey-helldar** [reply]
Essentially, packages separate translations to visually separate files.

In our case, all keys are translated in one place.

All Laravel packages that use the `__()` (alias of `trans()`) function will support translating the keys from our files.

I'll try to suggest adding our project to the official documentation, but I don't know how Taylor will decide.

**Reply by andrey-helldar** [reply]
I have submitted PR to Nova (https://github.com/laravel/nova-docs/pull/324) and Spark Classic (https://github.com/laravel/spark-docs/pull/69) projects.
I did not find a repository for [Spark Stripe](https://spark.laravel.com/docs/1.x/spark-stripe/customization.html#localization) documentation :(

**Reply by luisprmat** [reply]
And what if we create the **spark** and **nova** directories inside `locales/{xx}/spark/xx.json` and `locales/{xx}/nova/xx.json` for each language `xx`, with only translations for *spark* and *nova*? I like this idea because we can clean the `xx.json` of translations that will not be used if these paid packages are not installed. It is not good to depend on everything that Taylor decides

**Reply by caouecs** [reply]
yes, but we will have a same tag in two files

**Reply by andrey-helldar** [reply]
Taylor refused personally: https://github.com/laravel/nova-docs/pull/324 and https://github.com/laravel/spark-docs/pull/69

In my opinion, there is no need to create a separate folder. Personally, I like this structure better:

```
locales/fr/fr.json
locales/fr/nova.json
locales/fr/spark.json
locales/fr/auth.php
locales/fr/pagination.php
locales/fr/passwords.php
locales/fr/validation.php
locales/fr/validation-inline.php
```

I think this is the best method because we don't have a lot of files yet to separate them into folders.

Therefore, I propose to take it. What do you think?

**Reply by andrey-helldar** [reply]
Honestly, I expected a refusal. I'm not Spatie 😂

**Reply by luisprmat** [reply]
I'm agree with you @andrey-helldar

**Reply by luisprmat** [reply]
But now how to separate the files without losing the translations already completed ... humm. That takes a lot of work ...

**Reply by andrey-helldar** [reply]
I have an idea:

I need to take the original package localization files and I will write a service that will take only the necessary keys from our translated files.

These files will be saved for nova and spark, and the keys will be deleted from the main file.

**Reply by masterix21** [reply]
@andrey-helldar, there are two Spark translations:
- spark-stripe
- spark-paddle

I have both, but if you don't publish the `it.json`, `es.json`, etc files in the spark folder, works only with the English language.

Check here:
[Spark Stripe v1.1.3 - English](https://github.com/Laravel-Lang/lang/files/6437836/en.json.zip)
[Spark Paddle v1.1.0 - English](https://github.com/Laravel-Lang/lang/files/6437833/en.json.zip)

**Reply by andrey-helldar** [reply]
@masterix21, this is strange...

How does Spark access translation keys?

`__('Foo Bar')` or `__('<spark_key>::Foo Bar')`?

Because if, as in the first case, then the translation should automatically be applied, since the mechanics of Laravel are used.

Am I getting it right that Spark Stripe is [Spark Classic](https://spark-classic.laravel.com)?

**Reply by andrey-helldar** [reply]
@masterix21, @caouecs, @luisprmat, given that we plan to take out the Spark keys, do you think we can take out the rest of the keys as well?

In this case, we will get a lot of files that can be divided into folders.

I suggest a structure like this:

```
locales/xx/base.json
locales/xx/base/auth.php
locales/xx/base/pagination.php
locales/xx/base/passwords.php
locales/xx/base/validation.php
locales/xx/base/validation-inline.php

locales/xx/packages/cashier.json
locales/xx/packages/fortify.json
locales/xx/packages/jetstream.json
locales/xx/packages/nova.json
locales/xx/packages/spark-stripe.json
locales/xx/packages/spark-paddle.json
```

**Reply by luisprmat** [reply]
Apparently the only packages that require separate files are **spark** (and **nova**?). It seems to me that the base `xx.json` + the other packages can be left in one, also that they do not contain many keys by themselves. 
I would also not like to rename `xx.json` to `base.json`, I prefer the way it is copied to make it work.

```
locales/xx/xx.json
locales/xx/auth.php
locales/xx/pagination.php
locales/xx/passwords.php
locales/xx/validation.php
locales/xx/validation-inline.php

locales/xx/stripe/spark/xx.json
locales/xx/paddle/spark/xx.json
locales/xx/nova/xx.json
```

**Reply by andrey-helldar** [reply]
Personally, I don't like creating many folders so that each folder contains only one file.

BUT it looks nice :)

Now we have a problem that if one of the packages changes the name of the key, we will not be able to remove the extra one from the file, since we will not be sure that it is not used in other packages.

By splitting the translation into separate files, it will be easier for us to keep them up to date.

For example, Nova added 424 new keys to the [json file](https://github.com/Laravel-Lang/lang/discussions/1547#discussioncomment-474033).

As for copying files, in my opinion, it is better to use a translation manager - it will do everything by itself :)

For example: [andrey-helldar/laravel-lang-publisher](https://github.com/andrey-helldar/laravel-lang-publisher).

**Reply by andrey-helldar** [reply]
Guys...

![image](https://user-images.githubusercontent.com/10347617/117426888-196e9a00-af2d-11eb-9843-6c8a5f5d60f4.png)

Laravel Cashier Paddle: https://laravel.com/docs/8.x/cashier-paddle#introduction
Laravel Cashier Stripe: https://laravel.com/docs/8.x/billing#introduction

```
locales/xx/xx.json
locales/xx/auth.php
locales/xx/pagination.php
locales/xx/passwords.php
locales/xx/validation.php
locales/xx/validation-inline.php

locales/xx/stripe/spark/xx.json
locales/xx/stripe/cashier/xx.json
locales/xx/paddle/spark/xx.json
locales/xx/paddle/cashier/xx.json
locales/xx/nova/xx.json
```

?

**Reply by andrey-helldar** [reply]
Now I like this structure better:

```
locales/xx/xx.json
locales/xx/auth.php
locales/xx/pagination.php
locales/xx/passwords.php
locales/xx/validation.php
locales/xx/validation-inline.php

locales/xx/nova.json
locales/xx/stripe/spark.json
locales/xx/stripe/cashier.json
locales/xx/paddle/spark.json
locales/xx/paddle/cashier.json
```

**Reply by luisprmat** [reply]
If we look for translations into package **Laravel Cashier Paddle** these aren't found. I don't know if  they use the translations of another packages.

**Reply by andrey-helldar** [reply]
In some packages, Taylor did not carry translation keys into separate files, but simply uses the `__()` helper in the code.

**Reply by andrey-helldar** [reply]
@masterix21, hm...

![image](https://user-images.githubusercontent.com/10347617/118514013-dd151800-b73c-11eb-9118-5bdda04fe4f6.png)

**Reply by masterix21** [reply]
@andrey-helldar, see yourself...

<img width="739" alt="screen-1" src="https://user-images.githubusercontent.com/6555012/118526874-fcae3f80-b740-11eb-9a47-f28c52ec42b4.png">

<img width="741" alt="screen-2" src="https://user-images.githubusercontent.com/6555012/118526882-fe780300-b740-11eb-908d-605c2f884456.png">

**Reply by andrey-helldar** [reply]
@masterix21, Just an interesting point. Spark is a package for Laravel.
It uses json keys for translation. For example:

```
__('Billing Management')
```

This means that by placing the translation of this key in the main application's `/resources/lang/*.json` file, Laravel should automatically load it.

Said said that too, pointing to the location of the translation files in the app.

**Reply by andrey-helldar** [reply]
Said said 😂

**Reply by masterix21** [reply]
<img width="1734" alt="screen-1" src="https://user-images.githubusercontent.com/6555012/118562374-de5e3900-b76c-11eb-83f5-cdebf2f7759a.png">

<img width="1920" alt="screen-2" src="https://user-images.githubusercontent.com/6555012/118562398-e918ce00-b76c-11eb-8f5a-1bef6eaf2044.png">

<img width="1920" alt="screen-3" src="https://user-images.githubusercontent.com/6555012/118562412-eddd8200-b76c-11eb-9304-26fce6617daa.png">

I hope you will trust me, now :)

**Reply by andrey-helldar** [reply]
@masterix21, I believe you, but knowing how the translation system works in Laravel, I have doubts.

Do you know what I didn't see in your screenshots?

You have a `resources/lang/it.json` file. Does this file have a `Return to :appName` key?

Perhaps the translation does not go through precisely because of the lack of a key.

Can you also show how this key is called in the code?

**Reply by masterix21** [reply]
> Does this file have a Return to: appName key?

Yes, I'm using the `laravel-lang/lang` latest version.

> Can you also show how this key is called in the code?

Yes. The billing portal uses VueJS, and the translations are taken from a Javascript `window.translations` variable. Apparently they use `{{ __('Billing Management') }}`, but the syntax is VueJS (not blade).

Javascript variable data comes from:
![image](https://user-images.githubusercontent.com/6555012/118618392-dd0c2b00-b7c3-11eb-80c8-f2914735228e.png)

**Reply by andrey-helldar** [reply]
@masterix21, I understood what the problem is.

For the backend, it makes no difference where the translation files are - Laravel will always contain them in one object, and in order for VueJs to receive the translation for the frontend, it uses json content insertion.

Therefore, Spark does not display the translation.

I am currently [working](https://github.com/andrey-helldar/laravel-lang-publisher/pull/103) on improving my package, which will save Spark translation keys in the correct place :)

But it will only work with the new framework being developed [here](https://github.com/Laravel-Lang/lang/pull/1748).

**Reply by luisprmat** [reply]
@masterix21 with the new _split packages feature_ you only will need copy the file `locales/it/packages/spark-stripe.json` (or `spark-paddle`) and rename it as `resources/lang/spark/it.json`. That should work!

**Reply by andrey-helldar** [reply]
@masterix21, when we release version 10 (https://github.com/Laravel-Lang/lang/pull/1748#issuecomment-842980114), you can install a [translation manager](https://github.com/andrey-helldar/laravel-lang-publisher) in your project.

I [made](https://github.com/andrey-helldar/laravel-lang-publisher/pull/103) the code flexible. It can [detect](https://github.com/andrey-helldar/laravel-lang-publisher/tree/features/split-packages/src/Plugins) which packages are installed in the application and install only the necessary translation files.

```
php artisan lang:update

php artisan lang:add fr it de

php artisan lang:rm fr it de
```

![image](https://user-images.githubusercontent.com/10347617/118717346-a0c1e480-b82e-11eb-9a5b-97a7c1cdce8e.png)

**Reply by andrey-helldar** [reply]
@masterix21, can you tell me the names of the directories where spark is installed?

```
vendor/laravel/spark
vendor/laravel/spark-paddle
```
?

I only [found](https://spark.laravel.com/docs/1.x/installation.html#installing-spark-via-composer) spark paddle.

![image](https://user-images.githubusercontent.com/10347617/118942214-30fd4800-b95b-11eb-9d09-414af79e8d98.png)

**Reply by andrey-helldar** [reply]
> Apparently the only packages that require separate files are **spark** (and **nova**?). It seems to me that the base `xx.json` + the other packages can be left in one, also that they do not contain many keys by themselves.
> I would also not like to rename `xx.json` to `base.json`, I prefer the way it is copied to make it work.
> 
> ```
> locales/xx/xx.json
> locales/xx/auth.php
> locales/xx/pagination.php
> locales/xx/passwords.php
> locales/xx/validation.php
> locales/xx/validation-inline.php
> 
> locales/xx/stripe/spark/xx.json
> locales/xx/paddle/spark/xx.json
> locales/xx/nova/xx.json
> ```

I tried this architecture and it makes it difficult to access files because there are too many subfolders.

So I came up with a structure like this:

```
locales/xx/xx.json
locales/xx/auth.php
locales/xx/pagination.php
locales/xx/passwords.php
locales/xx/validation.php
locales/xx/validation-inline.php

locales/xx/packages/cashier.json
locales/xx/packages/fortify.json
locales/xx/packages/jetstream.json
locales/xx/packages/nova.json
locales/xx/packages/spark-paddle.json
locales/xx/packages/spark-stripe.json
```

**Reply by masterix21** [reply]
> @masterix21, can you tell me the names of the directories where spark is installed?
> 
> ```
> vendor/laravel/spark
> vendor/laravel/spark-paddle
> ```
> 
> ?
> 
> I only [found](https://spark.laravel.com/docs/1.x/installation.html#installing-spark-via-composer) spark paddle.
> 
> ![image](https://user-images.githubusercontent.com/10347617/118942214-30fd4800-b95b-11eb-9d09-414af79e8d98.png)

Using spark-paddle you will have `vendor/laravel/spark-paddle`

Using spark-stripe you will have `vendor/laravel/spark-stripe`

**Reply by andrey-helldar** [reply]
@masterix21, thanks!

**Reply by andrey-helldar** [reply]
@masterix21, you can use translation manager :)

It's released: [v10.0.0](https://github.com/andrey-helldar/laravel-lang-publisher/releases/latest)

**Reply by masterix21** [reply]
🥳🥳🥳

**Reply by luisprmat** [reply]
@andrey-helldar @caouecs One question, in this new version, what should we do as contributors if there are new english translation keys? Add them to the `source` directory files? Or is there a Github actions script that updates the packages and searches for them automatically?

**Reply by andrey-helldar** [reply]
@luisprmat, 

Packages that do not have translation files, but use them:

- laravel/fortify
- laravel/jetstream

These packages are distributed under the MIT license and auto search for keys is made for them. They automatically generate files [source/packages/fortify.json](https://github.com/Laravel-Lang/lang/blob/master/source/packages/fortify.json) and [source/packages/jetstream.json](https://github.com/Laravel-Lang/lang/blob/master/source/packages/jetstream.json) and keep them up to date.

For all other packages, there are translation files in the projects, so we will just copy the new files.

The same is with paid packages (Spark, Nova).

**Reply by andrey-helldar** [reply]
For example: https://github.com/Laravel-Lang/lang/pull/1755
